### PR TITLE
Feature/icds 1433 reprocess alerts

### PR DESCRIPTION
**Pull request by adil-uddin**
https://dimagi-dev.atlassian.net/projects/ICDS/issues/ICDS-1433

##### SUMMARY

Presently in HQ when we update any condition in the SMS rule, the rule is processed again. However, if we only update the content-message in the SMS rule the rule does not need to be processed again because the content is injected only at the end while sending the SMS. 
Avoiding the processing, would save a lot of processing which is wasted in the cases when only the content of the SMS is updated.
The same thing is achieved via bulk upload for conditional alerts but that needs a bulk upload each time even if a single alert needs to be changed, so having it in UI makes it easier and quicker.

### CODE REVIEW

**Comment by adil-uddin** [comment]
@mkangia let me know next steps

File: corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js [hunk 1]
```diff
@@ -34,5 +34,29 @@ hqDefine("scheduling/js/conditional_alert_main", [
...
+        $('.main-form').on( "submit", function( event ) {
+            var finalFormContent = {};
+            _.each(
+                $('.main-form').serializeArray(),
+                function(obj, index) {finalFormContent[obj.name] = obj.value}
+            );
+            // compare changes with initialFormContent to check if only 'message' has changed
+            // to convey to HQ to reprocess or not
+            var formContentChanges = []
+            Object.keys(finalFormContent).forEach(key => {
```
[the remark this answers is not in the document]
Comment (adil-uddin) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:54]: @mkangia is the js code ES6 supported ?

[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:54]: No, i don't think you can do that.

[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:54]: I think HQ will be okay with it, but not stickler (clearly). I'm not sure if the UglifyJS will minify it correctly before deploy. It's most common in this codebase to use underscore for looping, as you're doing above.

File: corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js [hunk 2]
```diff
@@ -34,5 +34,29 @@ hqDefine("scheduling/js/conditional_alert_main", [
         // setup tab
         basicInformation.setRuleTabVisibility();
         $("#conditional-alert-basic-info-panel").koApplyBindings(basicInformation);
+
+
+        var initialFormContent = {};
```
Comment (orangejenny) [inline corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:39]: This is a stylistic comment, so feel free to ignore, but it'd be more concise to do these object creations each as a single statement, so this would become:
```
var initialFormContent = _.object(_.map($('.main-form').serializeArray(), function (obj) {
    return [obj.name, obj.value];
}));
```

Comment (adil-uddin) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:39]: Thanks @orangejenny  this is great, though I'm not sure `_` here is `underscore.js` or `lodash`. It seems `_.object` is removed in lodash with `_.fromPairs`.

Comment (orangejenny) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:39]: It's underscore. We do have lodash in `package.json` but only a dependency of other libraries.

I was going to say you can check the dependencies in the `hqDefine` call at the top of the file...and then realized it wasn't there. Can you add it? I'm surprised this page doesn't break on staging/prod (it won't break locally because local requirejs config is more relaxed than prod), underscore must make itself available as a global. But we should be in the habit of declaring all dependencies explicitly.

Comment (adil-uddin) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js]: done

File: corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js [hunk 3]
```diff
@@ -51,19 +51,26 @@ hqDefine(
...
-                }
+                var runMessagingRule = !_.isEqual(
+                    ["message"],
+                    formContentChanges
+                );
+                // to convey to HQ to reprocess or not
+                var input = $("<input />")
+                    .attr("type", "hidden")
+                    .attr("name", "runMessagingRule")
+                    .attr("value", runMessagingRule.toString());
```
Comment (orangejenny) [inline corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:71]: nitpick: this will be a string by the time it gets to python even without the explicit cast, right?

File: corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js [hunk 4]
```diff
@@ -3,6 +3,7 @@ hqDefine(
     [
         "jquery",
         "knockout",
+        "underscore",
```
Comment (orangejenny) [inline corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:6]: You also need to add `_` as the third parameter in the function call: each dependency is matched up with the corresponding parameter, which is then used in the function body to access that dependency. Dependencies that aren't directly referenced don't need a parameter.

I'd expect this code to fail when it references `initialPageData`, which is correspond to underscore now instead of the initial page data module.

File: corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js [hunk 5]
```diff
@@ -1,62 +1,69 @@
...
-], function ($, ko, initialPageData) {
-    function basicInformationTab(name) {
-        var self = {};
-        self.name = ko.observable(name);
-        self.basicTabValid = ko.computed(function () {
-            return !_.isEmpty(self.name().trim());
-        });
-        self.setRuleTabVisibility = function () {
-            if (self.basicTabValid()) {
+hqDefine(
```
Comment (orangejenny) [inline corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:1]: It's helpful to reviewers to separate formatting-only changes from functionality changes; it's hard to tell what really changed in this commit.

**Comment by orangejenny** [comment]
The js dependency needs to be updated to add `_`, but otherwise this looks good.

File: corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js (hunk 3 above)
Comment (adil-uddin) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js]: true, seems we don't need to explicitly call toString. Removing.

File: corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js (hunk 4 above)
Comment (adil-uddin) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:6]: Thanks. Added _ as third param

File: corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js (hunk 5 above)
Comment (adil-uddin) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:1]: Thanks Jenny, will keep this in mind next time 👍

**Comment by adil-uddin** [comment]
Updated as per the comments @orangejenny

**Review by orangejenny — APPROVED** [review]

**Comment by orangejenny** [comment]
Build failures look related to ICDS code splitting. Merging master might address them.

**Comment by adil-uddin** [comment]
Merged master the builds still fails due to some config issues @orangejenny

File: corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js [hunk 6]
```diff
@@ -1,38 +1,77 @@
...
+                var runMessagingRule = !_.isEqual(
+                    ["message"],
+                    formContentChanges
+                );
+                // to convey to HQ to reprocess or not
+                var input = $("<input />")
+                    .attr("type", "hidden")
+                    .attr("name", "runMessagingRule")
+                    .attr("value", runMessagingRule);
+                $(this).append(input);
```
Comment (mkangia) [inline corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:72]: can we add this in the actual django form instead and just set it to False in case the changes are just for "message"?
I say so because this is injecting some magic via javascript which i think can be difficult for devs to notice. With a hidden input field in django form and a help text or a comment in code, would make it explicit on why that exists.

File: corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js [hunk 7]
```diff
@@ -1,38 +1,77 @@
...
-});
+            var initialFormContent = getFormContents();
+
+            $(".main-form").on("submit", function () {
+                var finalFormContent = getFormContents();
+                // compare changes with initialFormContent to check if only 'message' has changed
+                var formContentChanges = [];
+                Object.keys(finalFormContent).forEach(function (key) {
+                    finalFormContent[key] !== initialFormContent[key]
+                        ? formContentChanges.push(key)
```
Comment (mkangia) [inline corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:60]: woah! that was a little hard to read, is that the standard way to break down ternary operator?

Comment (AmitPhulera) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:60]: Hey @mkangia by breaking down, do you mean that we should change it to if-else or something else?

Comment (mkangia) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:60]: Was mainly the indentation. I am fine with how it is, or if it can be made into if-else.

File: corehq/messaging/scheduling/forms.py [hunk 8]
```diff
@@ -184,6 +184,10 @@ class ContentForm(Form):
         label=ugettext_lazy("Intervals"),
     )
 
+    should_run_rule = CharField(
+        widget=HiddenInput,
+    )
```
Comment (mkangia) [inline corehq/messaging/scheduling/forms.py:189]: can we set default value to be "true" here itself?

File: corehq/messaging/scheduling/forms.py [hunk 9]
```diff
@@ -471,7 +479,7 @@ def compute_initial(domain, content):
             result['sms_callback_intervals'] = ', '.join(str(i) for i in content.reminder_intervals)
         else:
             raise TypeError("Unexpected content type: %s" % type(content))
-
+        result['should_run_rule'] = "true"
```
Comment (mkangia) [inline corehq/messaging/scheduling/forms.py:482]: why is this needed?

File: corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js [hunk 10]
```diff
@@ -56,20 +56,16 @@ hqDefine(
...
+                    ["content-message"],
                     formContentChanges
                 );
                 // to convey to HQ to reprocess or not
-                var input = $("<input />")
-                    .attr("type", "hidden")
-                    .attr("name", "runMessagingRule")
-                    .attr("value", runMessagingRule);
-                $(this).append(input);
+                $("#should_run_rule").val(runMessagingRule);
```
Comment (mkangia) [inline corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:68]: :+1:

Comment (mkangia) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:68]: also can this be
```
if (_.isEqual(["content-message"], formContentChanges)) {
    $("#should_run_rule").val(false)
}
```
else it would stay as default true

File: corehq/messaging/scheduling/forms.py [hunk 11]
```diff
@@ -422,6 +426,10 @@ def get_layout_fields(self):
                 crispy.Field('sms_callback_intervals'),
                 data_bind="visible: $root.content() === '%s'" % ScheduleForm.CONTENT_SMS_CALLBACK,
             ),
+            crispy.Div(
+                crispy.Field('should_run_rule', id="should_run_rule"),
+                data_bind="value: should_run_rule",
```
Comment (mkangia) [inline corehq/messaging/scheduling/forms.py:431]: i don't think you need data_bind here since we are setting the value directly and not via Knockout model

File: corehq/messaging/scheduling/forms.py [hunk 12]
```diff
@@ -184,6 +184,10 @@ class ContentForm(Form):
         label=ugettext_lazy("Intervals"),
     )
 
+    should_run_rule = CharField(
```
Comment (mkangia) [inline corehq/messaging/scheduling/forms.py:187]: nit: `initiate_rule`

```
            if initiate_rule:
                initiate_messaging_rule_run(rule)
```

**Review by mkangia — COMMENTED** [review]
Looking good. Just some minor improvements and one query.

File: corehq/messaging/scheduling/forms.py (hunk 9 above)
Comment (AmitPhulera) [reply corehq/messaging/scheduling/forms.py:482]: It's to initialize the value to "true". I guess it can be achieved by default value also.

File: corehq/messaging/scheduling/forms.py (hunk 11 above)
Comment (AmitPhulera) [reply corehq/messaging/scheduling/forms.py:431]: Yeah. I started out by thinking of binding to the KO model but changed the approach later and then I forgot to remove it from code.

File: corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js [hunk 13]
```diff
@@ -1,38 +1,71 @@
...
+                // compare changes with initialFormContent to check if only 'message' has changed
+                var formContentChanges = [];
+                Object.keys(finalFormContent).forEach(function (key) {
+                    if (finalFormContent[key] !== initialFormContent[key]) {
+                        formContentChanges.push(key);
+                    }
+                });
+                if (!_.isEqual(["content-message"],formContentChanges)) {
+                    // to convey to HQ to reprocess or not
+                    $("#initiate_rule").val(false);
```
Comment (mkangia) [inline corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:65]: nit: is there a way we can handle this in the loop itself?
As soon as we find a key that is not in ['content-message'], we know that the rule needs to be initiated and can just set it to true.

Comment (AmitPhulera) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:65]: I think we will need to change `forEach` to `for` or `for` loop because `forEach` can't be break' ed in between.
In terms of readability, I am not sure which approach would be good.
@mkangia what would be your comments?

Comment (orangejenny) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:65]: [_.find](https://underscorejs.org/#find) can be a readable way to set up a loop that breaks on a condition.

**Comment by AmitPhulera** [comment]
Tagging @dimagi/product  to get inputs on if this can be a product-level change or should we keep it behind a feature flag?

File: corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js [hunk 14]
```diff
@@ -54,16 +54,13 @@ hqDefine(
...
+                var shouldInitiateRule = _.find(_.keys(finalFormContent), function (key) {
+                    if (finalFormContent[key] !== initialFormContent[key] && key !== "content-message") {
+                        return true;
                     }
                 });
-                if (!_.isEqual(["content-message"],formContentChanges)) {
-                    // to convey to HQ to reprocess or not
-                    $("#initiate_rule").val(false);
-                }
+                $("#initiate_rule").val(Boolean(shouldInitiateRule));
```
Comment (mkangia) [inline corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:62]: why do we need the `Boolean` call here?

Comment (AmitPhulera) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:62]: Because we wanted the end result to be `true` or `false`. 
`_.find` returns the key for which the condition is true or it will return `undefined` if the condition does not match for any of the keys.

Comment (mkangia) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:62]: cool, is that same as doing `!!shouldInitiateRule`?
(not saying you should change it to that, just understanding if this does something different)

Comment (AmitPhulera) [reply corehq/messaging/scheduling/static/scheduling/js/conditional_alert_main.js:62]: Yes. It's the same thing.

**Review by mkangia — APPROVED** [review]

**Comment by mkangia** [comment]
@AmitPhulera can you confirm if we need to avoid retrigger in case the alert is inactive as well? Ticket description says that.

**Comment by mkangia** [comment]
Closing this because its not needed anymore at least from the ICDS perspective. So if feels like a good product change, can be then reopened.
